agent/provider/src/rules.rs

```diff
@@ -199,105 +202,168 @@ impl RulesManager {
-    ) -> CheckRulesResult {
-        let cfg = self.rulestore.config.read().unwrap();
+    ) -> Result<()> {
+        if let Some(props) = manifest_sig {
+            //TODO Add verification of permission tree when they will be included in x509 (as there will be in both Rules)
+            self.verify_permissions(&props.cert, demand_permissions_present)
+                .map_err(|e| anyhow!("Audited-Payload rule: {e}"))?;
 
-        if cfg.outbound.enabled.not() {
-            log::trace!("Outbound is disabled.");
+            self.keystore
+                .verify_signature(props.cert, props.sig, props.sig_alg, props.manifest_encoded)
+                .map_err(|e| anyhow!("Audited-Payload rule: {e}"))?;
 
-            return CheckRulesResult::Reject("outbound is disabled".into());
+            let mode = &self
+                .rulestore
+                .config
+                .read()
+                .unwrap()
+                .outbound
+                .audited_payload
+                .default
+                .mode;
+
+            self.check_mode(mode, manifest)
+                .map_err(|e| anyhow!("Audited-Payload {e}"))
+        } else {
+            Err(anyhow!("Audited-Payload rule requires manifest signature"))
         }
+    }
 
-        match cfg.outbound.everyone {
-            Mode::All => {
-                log::trace!("Everyone is allowed for outbound");
+    fn check_partner_rule(
+        &self,
+        manifest: &AppManifest,
+        node_identity: Option<String>,
+        requestor_id: NodeId,
+    ) -> Result<()> {
+        let node_identity =
+            node_identity.ok_or_else(|| anyhow!("Partner rule requires partner certificate"))?;
+
+        let verified_cert = self
+            .keystore
+            .verify_golem_certificate(&node_identity)
+            .map_err(|e| anyhow!("Partner {e}"))?;
+
+        if requestor_id != verified_cert.node_id {
+            return Err(anyhow!(
+                "Partner rule nodes mismatch. requestor node_id: {} but cert node_id: {}",
+                requestor_id,
+                verified_cert.node_id
+            ));
+        }
 
-                return CheckRulesResult::Accept;
+        self::verify_golem_permissions(
+            &verified_cert.permissions,
+            &manifest.get_outbound_requested_urls(),
+        )
+        .map_err(|e| anyhow!("Partner {e}"))?;
+
+        let cert_ids = verified_cert
+            .cert_ids_chain
+            .iter()
+            .map(|i| i.hash.clone())
+            .collect::<Vec<_>>();
+
+        for cert_id in cert_ids.iter() {
+            if let Some(rule) = self
+                .rulestore
+                .config
+                .read()
+                .unwrap()
+                .outbound
+                .partner
+                .get(cert_id)
+            {
+                return self
+                    .check_mode(&rule.mode, manifest)
+                    .map_err(|e| anyhow!("Partner {e}"));
             }
```

Comment:
This is incorrect. Not all certificates provided by Requestor need to have permissions. Only the leaf certificate,
which directly signed `node.json`.

You can have for example Golem Root, which you trust only with whitelist, but other Golem certificates signed by root, which you trust more for some reason

Reply:
It is correct as discussed